accesslog: Pass destination identity from L7 proxies to access logs

jrajahalme opened the pull request:
Always pass both the source and destination security identities that were used
in L7 proxy policy enforcement to access logs and make sure that also
destination security identity is access logged in L7 messages in all cases. This
fixes cases where destination security identity may have been access logged
as 'WORLD' even though the destination is a pod in the cluster.

This also makes sure that the correct security identities are access logged,
even in cases where endpoint's security identity may have changed between
the time of the proxy event and log handling in the Cilium agent.

4th commit fixes an old discrepancy by passing source and destination
addressing info in access logs correctly instead of relying Hubble to swap
them around for responses. This has the benefit that also cilium monitor
output has correct security identity for L7 responses.

Note that this changes cilium proxy output that used to incorrectly show
security identities as "A->B" for both requests and responses. Now these
are formatted as "A->B" and "B->A" for requests and responses,
respectively.

Final commit avoids losing Envoy access log messages in the case the
local endpoint can not be found (any more).

This is a prerequisite for #18894

Fixes: #9558
Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

```release-note
Cilium monitor now correctly reports security identities for L7 flows.
```

jrajahalme committed:
accesslog: Pass destination identity from L7 proxies to access logs

Pass destination security identity, as used at the time of policy
enforcement, from L7 proxies to access logs. This can be 0 if not
resolved by the proxy (e.g., by an ingress proxy), but in those cases the
logger now resolves the destination IP to the corresponding security
identity even when the destination is not a local endpoint.

Unify EndpointInfoRegistry interface calls FillEndpointIdentityByID() and
FillEndpointIdentityByIP() into a new FillEndpointInfo() that takes both
IP and numeric identity as parameters. This is used for both source and
destination endpoint info in logger.Addressing() so that EndpointInfo no
longer needs to be overwritten in notifyOnDNSMsg() after the Addressing()
tag.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
proxy: Pass addressing info for response logs correctly

Pass source vs. destination addressing info in access logs correctly
instead of relying Hubble to swap them around for responses. This has the
benefit that also cilium monitor output has correct address and security
identity for L7 responses.

Note that this changes cilium monitor output that used to incorrectly show
security identities as "A->B" for both requests and responses. Now these
are formatted as "A->B" and "B->A" for requests and responses,
respectively.

Fixes: #9558
Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

michi-covalent reviewed (approved):
:heavy_check_mark:  reviewed `pkg/hubble`

glibsm reviewed (approved):
`pkg/hubble/parser/seven` looks good.

aditighag commented on pkg/proxy/logger/logger.go:
```diff
@@ -105,7 +99,7 @@ func NewLogRecord(endpointInfoRegistry EndpointInfoRegistry, localEndpointInfoSo
 // source.
 func (lr *LogRecord) fillEndpointInfo(info *accesslog.EndpointInfo, ip net.IP, secId identity.NumericIdentity) {
-	lr.endpointInfoRegistry.FillEndpointInfo(info, ip, secId)
+	endpointInfoRegistry.FillEndpointInfo(info, ip, secId)
```
Does this need to be a method receiver anymore given that `endpointInfoRegistry` is moved out of `LogRecord`.

jrajahalme replied:
No, will remove this function and use `endpointInfoRegistry` directly from the call sites, thanks!

aditighag commented on daemon/cmd/fqdn.go:
```diff
@@ -467,15 +467,22 @@ func (d *Daemon) notifyOnDNSMsg(lookupTime time.Time, ep *endpoint.Endpoint, epI
+		addrInfo.SrcIdentity = ep.GetIdentity()
+		addrInfo.DstIPPort = serverAddr
+		addrInfo.DstIdentity = serverID
+	}
```
Can we move this block that populates various fields earlier in the function (L440) where it gives context for how the direction is interpreted - "The observation point is always Egress, however."? We can move the ep nil check condition before L434.

jrajahalme replied:
Nice, this reads now:
```
	// We determine the direction based on the DNS packet. The observation
	// point is always Egress, however.
	var flowType accesslog.FlowType
	var addrInfo logger.AddressingInfo
	if msg.Response {
		flowType = accesslog.TypeResponse
		addrInfo.DstIPPort = epIPPort
		addrInfo.DstIdentity = ep.GetIdentity()
		addrInfo.SrcIPPort = serverAddr
		addrInfo.SrcIdentity = serverID
	} else {
		flowType = accesslog.TypeRequest
		addrInfo.SrcIPPort = epIPPort
		addrInfo.SrcIdentity = ep.GetIdentity()
		addrInfo.DstIPPort = serverAddr
		addrInfo.DstIdentity = serverID
	}
```

aditighag commented on pkg/endpoint/endpoint.go:
```diff
@@ -2248,21 +2248,6 @@ func (e *Endpoint) Delete(conf DeleteConfig) []error {
-// GetProxyInfoByFields returns the ID, IPv4 address, IPv6 address, labels,
-// SHA of labels, and identity of the endpoint. Returns an error if the endpoint
-// is in the process of being deleted / has been deleted.
-func (e *Endpoint) GetProxyInfoByFields() (uint64, string, string, []string, string, uint64, error) {
```
How's this change related to the commit description?

jrajahalme replied:
Dead code after `LogRecord.localEndpointInfo` was removed. I split this change to a separate commit to make this clearer.

aditighag commented on pkg/envoy/accesslog_server.go:
```diff
@@ -164,16 +162,26 @@ func logRecord(endpointInfoRegistry logger.EndpointInfoRegistry, localEndpoint l
+	// swap them here so that the logs have the correct info.
+	// TODO (jrajahalme): Consider doing this at our Envoy filters instead?
+	var addrInfo logger.AddressingInfo
+	if flowType == accesslog.TypeResponse {
```
This warrants a comment saying how the direction is interpreted (similar to above).

jrajahalme replied:
I changed the comment to this to make this clearer:
```
	// Response access logs from Envoy inherit the source/destination info from the request log
	// message. Swap source/destination info here for the response logs so that they are
	// correct.
	// TODO (jrajahalme): Consider doing this at our Envoy filters instead?
```

aditighag commented on pkg/proxy/logger/logger.go:
```diff
@@ -161,9 +148,7 @@ func (logTags) Addressing(i AddressingInfo) LogTag {
-				if lr.ObservationPoint == accesslog.Ingress {
-					lr.fillEndpointInfo(&lr.SourceEndpoint, ip, i.SrcIdentity)
-				}
+				lr.fillEndpointInfo(&lr.SourceEndpoint, ip, i.SrcIdentity)
```
See comment @L102.

jrajahalme replied:
Done :-)

aditighag commented on pkg/proxy/logger/logger.go:
```diff
@@ -102,69 +102,10 @@ func NewLogRecord(endpointInfoRegistry EndpointInfoRegistry, localEndpointInfoSo
+// fillEndpointInfo fills the EndpointInfo fields using identity sent by
+// source.
+func (lr *LogRecord) fillEndpointInfo(info *accesslog.EndpointInfo, ip net.IP, secId identity.NumericIdentity) {
+	lr.endpointInfoRegistry.FillEndpointInfo(info, ip, secId)
```
This is again modified in the next commit. Can we pull common changes in a separate commit?

jrajahalme replied:
I pulled this out from the last commit, which is now a lot more self contained, thanks!

aditighag reviewed (changes requested):
While the commit descriptions mention the issues that are being fixed, I didn't fully understand the implementation details. 

>Remove redundant code from pkg/fqdn and simplify pkg/proxy/logger now
that the source and destination endpoint info is filled in the same way.

Can the clean-up related changes be moved to separate commits so that reviewers know the relevant changes to review?

jrajahalme committed:
logger: Use FillEndpointInfo regardless the observation point

Fill source and destination endpoint info in LogRecord
using the same FillEndpointInfo() function, making
LogRecord.localEndpointInfo related functions obsolete.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
logger: Move endpointInfoRegistry out of LogRecord

endpointInfoRegistry is essentially a package level variable, as it is
always the same. Add a new seetter to set it once in the beginning.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
envoy: Log access logs even when local endpoint does not exist

Do not discard access logs if the local endpoint can not be found. This
can happen if endpoint is removed before all access logs for it are
received from Envoy.  Endpoint's proxy stats still can not be updated in
that case.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

aditighag commented on pkg/proxy/logger/logger.go:
```diff
@@ -50,18 +50,19 @@ const (
+
+var endpointInfoRegistry EndpointInfoRegistry
 
-	// endpointInfoRegistry provides access to any endpoint's information given
```
The context for the variable is now lost.

jrajahalme replied:
Thanks for the review! I restored the comment, since I had to rebase to master to pick up CI fixes anyway :-)

aditighag reviewed (approved):
Thanks for the revisions - looks much cleaner now. Doesn't need my follow-up review for the latest review comment.
